### PR TITLE
Refactor bootstrap to make use of go-git-providers

### LINKED ISSUE #799: Write instructions for bootstrapping using a writable git repository, for automation
If you just use `flux bootstrap`, you get a read-only key for the git repository. Many people will run image update automation in the same place as syncing, and it would be convenient for them to be able to use a single git repository both for syncing and for automation -- which would need it to be writable.

### LINKED ISSUE #835: [Gitlab] flux bootstrap fails for personal projects if they already exist
When I try to update my flux v2 installation using the bootstrap command it fails with an error from gitlab:

> failed to create project, error: POST https://gitlab.com/api/v4/projects: 400

I use the below bootstrap command to install / update flux v2, which worked until now:

```bash
ubuntu@srv1:~$ cat install-flux.sh 
curl -s https://toolkit.fluxcd.io/install.sh | sudo bash

export GITLAB_TOKEN=???????????????

flux bootstrap gitlab \
  --owner=isnull \
  --repository=myrepo \
  --branch=master \
  --path=k8 \
  --token-auth \
  --personal
```

Executing the flux bootstrap yields the error:
```bash
ubuntu@srv1:~$ sh install-flux.sh 
[INFO]  Downloading metadata https://api.github.com/repos/fluxcd/flux2/releases/latest
[INFO]  Using 0.7.4 as release
[INFO]  Downloading hash https://github.com/fluxcd/flux2/releases/download/v0.7.4/flux_0.7.4_checksums.txt
[INFO]  Downloading binary https://github.com/fluxcd/flux2/releases/download/v0.7.4/flux_0.7.4_linux_amd64.tar.gz
[INFO]  Verifying binary download
[INFO]  Installing flux to /usr/local/bin/flux
► connecting to gitlab.com
✗ failed to create project, error: POST https://gitlab.com/api/v4/projects: 400 {message: {limit_reached: []}, {name: [has already been taken]}, {path: [has already been taken]}}
```

Sys info:
```bash
ubuntu@srv1:~$ flux check
► checking prerequisites
✔ kubectl 1.20.2 >=1.18.0
✔ Kubernetes 1.19.5-34+8af48932a5ef06 >=1.16.0
► checking controllers
✔ source-controller is healthy
► ghcr.io/fluxcd/source-controller:v0.5.6
✔ kustomize-controller is healthy
► ghcr.io/fluxcd/kustomize-controller:v0.5.3
✔ helm-controller is healthy
► ghcr.io/fluxcd/helm-controller:v0.4.4
✔ notification-controller is healthy
► ghcr.io/fluxcd/notification-controller:v0.5.0
✔ all checks passed
```

Maybe some Gitlab project api change caused this?

### LINKED ISSUE #916: Bootstrap picks existing repo instead of creating new one
### Describe the bug

When a repository already exists with a name `foo-bar` and you attempt to bootstrap with `--repository=foo` it instead chooses the already-existing `foo-bar`. I don't know if the hyphen has anything to do with it or not, I haven't looked into it further.

### To Reproduce

1. Create a repository called `foo-bar` on Github
2. Run `flux bootstrap github --repository=foo`
3. Observe there's no new repository created, but your manifests have been committed to the existing repository.

### Expected behavior

A new repository called `foo` would be created.

### Additional context

- Kubernetes version: 1.8
- Git provider: Github

Below please provide the output of the following commands:

```cli
flux --version
flux version 0.7.7
```

### LINKED ISSUE #924: GitLab bootstrap - git clone error: repository not found
### Describe the bug

I am using a paid gitlab.com. The site reports this version: GitLab Enterprise Edition 13.9.0-pre 04013f0902b.
I am using flux2 v0.8.0.

I have a private group called 'company-de'. All subgroups are also private. I have a subgroup called 'project' and another subgroup in it called 'ops'.
The url for the repo i am trying to create would be: https://gitlab.com/company-de/project/ops/flux-system.

I am NOT trying to bootstrap flux for a personal repo.
This problem happens in a windows terminal, git terminal for windows (git 2.30.0.windows.2), ubuntu 20.04 LTS on Windows WSL2 (git 2.25.1)

1)
When I provide owner=company-de/project/ops i get this error:
```cli
► connecting to gitlab.com
✗ failed to get project, error: failed to find group named 'company-de'
```

2)
When I provide owner=project/ops i get the following:
```cli
► connecting to gitlab.com
✔ repository created
✗ git clone error: repository not found
```

NOTE: in the second case the repository is created. Branch 'master' contains a readme file.

I tried all the possible combinations I can think of and I am stuck on the git clone error. I tried to use --verbose but that doesnt help. It would be nice to get better error logs. I wanted to see what is the command that flux executes to clone the repo.

I changed the repo name so it doesnt contain a dash. I tried to bootstrap one level up in 'project' (so its not a subgroup in a subgroup). It didnt help. I dont see what is the problem.

### To Reproduce

Steps to reproduce the behaviour:

export GITLAB_TOKEN=x
(I went to my user account details in gitlab, then to PAT section, then i created a new token with api permissions. I also tried adding all possible permissions for PAT)
(i have a git private key available on my system too, for ssh - git clone works when i do it manually)

1)
```cli
flux bootstrap gitlab \
  --owner=project/ops \
  --repository=flux-system \
```

2)
```cli
flux bootstrap gitlab \
  --ssh-hostname=gitlab.com \
  --owner=project/ops \
  --repository=flux-system \
```

I also tried --token-auth, --private, --path, --branch, --components but these params dont change anything.

### Expected behavior

I expect the bootstrap process to finish.

### Additional context

Below please provide the output of the following commands:

```cli
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-19T11:45:27Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

$ git --version
git version 2.25.1

$ flux --version
flux version 0.8.0

$ flux check
► checking prerequisites
✔ kubectl 1.20.1 >=1.18.0
✔ Kubernetes 1.20.2 >=1.16.0
► checking controllers
✗ source-controller: deployment not found
✗ kustomize-controller: deployment not found
✗ helm-controller: deployment not found
✗ notification-controller: deployment not found
```

### LINKED ISSUE #1021: Bootstrap creates empty commits
Repeatedly running bootstrap creates empty commits for me.

```
% git log --all --reverse --format='%h %t %cn %s' | grep v0.9.0
8606dfe 9d771ef flux Add flux v0.9.0 components manifests
f13d20f 9d771ef flux Add flux v0.9.0 sync manifests
7db09ce cc63d62 flux Add flux v0.9.0 components manifests
36744d1 e6f084e flux Add flux v0.9.0 sync manifests
7cfbf52 e6f084e flux Add flux v0.9.0 components manifests
2253d60 e6f084e flux Add flux v0.9.0 sync manifests
9d88c6b e6f084e flux Add flux v0.9.0 components manifests
e2d2afa e6f084e flux Add flux v0.9.0 sync manifests
8999512 e6f084e flux Add flux v0.9.0 components manifests
bf106c0 e6f084e flux Add flux v0.9.0 sync manifests
e8f0b79 e6f084e flux Add flux v0.9.0 components manifests
b7a0372 e6f084e flux Add flux v0.9.0 sync manifests
742eacf e6f084e flux Add flux v0.9.0 components manifests
0d0743f e6f084e flux Add flux v0.9.0 sync manifests
e9e5e24 e6f084e flux Add flux v0.9.0 components manifests
fb544b6 e6f084e flux Add flux v0.9.0 sync manifests
37a5419 e6f084e flux Add flux v0.9.0 components manifests
3a4bcc8 e6f084e flux Add flux v0.9.0 sync manifests
0ffcdfe 60100dd flux Add flux v0.9.0 components manifests
e5764b6 60100dd flux Add flux v0.9.0 sync manifests

% git show e5764b6
commit e5764b6207e129b80bb9629e9b07e541ebbe785a (HEAD -> main, origin/main, origin/HEAD)
Author: flux <jbergler@users.noreply.gitlab.com>
Date:   Tue Mar 2 19:59:34 2021 +1300

    Add flux v0.9.0 sync manifests
```
### To Reproduce

Steps to reproduce the behaviour:

1. Install flux 0.9.0
2. Run bootstrap (against a gitlab repo, although I don't think it matters)
3. Nuke/recreate my k8s cluster
4. Run bootstrap again

### Expected behavior

The first time bootstrap should commit the manifest (or changes to them if the versions changed), subsequently I would expect the commit to be skipped (per https://github.com/fluxcd/flux2/blob/main/cmd/flux/bootstrap_gitlab.go#L287)

### Additional context

- Kubernetes version: 1.19
- Git provider: gitlab

```cli
% flux --version

flux version 0.9.0

% flux check

► checking prerequisites
✔ kubectl 1.20.4 >=1.18.0-0
✔ Kubernetes 1.19.5+k3s2 >=1.16.0-0
► checking controllers
✔ helm-controller: healthy
► ghcr.io/fluxcd/helm-controller:v0.8.0
✔ kustomize-controller: healthy
► ghcr.io/fluxcd/kustomize-controller:v0.9.1
✔ notification-controller: healthy
► ghcr.io/fluxcd/notification-controller:v0.9.0
✔ source-controller: healthy
► ghcr.io/fluxcd/source-controller:v0.9.0
✔ all checks passed
```

### PR DESCRIPTION
**Pull request by hiddeco**
This pull request changes the implementation used to connect to Git
providers like GitHub from `fluxcd/pkg/git` to `fluxcd/go-git-providers`,
eliminating quite some bugs that were present in the GitLab integration.

In addition to this, the bootstrap logic has been factored out into an
(for now) internal package with two bootstrap "runners" `PlainGitBootstrapper`
and `GitProviderBootstrapper`. The `GitProviderBootstrapper` is a superset
of the `PlainGitBootstrapper` that besides `Reconciler` also implements the
`RepositoryReconciler`.

The Git actions rely on an interface, making it easier to support other
implementations than `go-git` at a later moment, to for example support
bootstrapping to Git servers that only support the v2 protocol.

By doing this, it was possible to introduce a new `flux bootstrap git`
command that bootstraps Flux to a plain Git repository and applies the
resources to the cluster, prompting the user to upsert the public key
by hand if a private key was generated.

As a finishing touch, almost all options are now configurable like they
already were for the `flux <install|create>` commands, making it possible
to have much finer grain configurations like e.g. the Git author that
makes the commit, the type of key that is generated, adding a `[skip ci]`
appendix to the commit message, and many more.

### Future enhancement possibilities:

- Assign different rights than `maintainer` to GitHub and GitLab teams
- Provide another Git implementation so bootstrapping to e.g. Azure DevOps is possible
- ...

### Help testing this PR :superhero:

1. Clone the repository and checkout the branch:

   ```console
   $ git clone https://github.com/fluxcd/flux2.git
   $ cd flux2
   $ git checkout -b go-git-providers-bootstrap origin/go-git-providers-bootstrap
   ```

2. Build the binary (requires Go 1.16 and Kustomize):

   ```console
   $ make build
   ```

3. Run one of the bootstrap commands:

   ```console
   $ ./bin/flux bootstrap git --help
   $ ./bin/flux bootstrap gitlab --help
   $ ./bin/flux bootstrap github --help
   ```

### Ref list

Fixes #799 
Fixes #835 
Fixes #916
Fixes #924
Fixes #1021
Fixes #1129
Fixes #1143

Addresses part of #43

### CODE REVIEW

File: pkg/bootstrap/git/gogit/gogit.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+		return false, err
+	}
+	return status.IsClean(), nil
+}
+
+func (g *goGit) Path() string {
+	return g.path
+}
+
+func isCouldntFindRemoteRefErr(err error, ref string) bool {
```
Comment (stefanprodan) [inline pkg/bootstrap/git/gogit/gogit.go:196]: maybe `isRemoteNotFoundErr`?

Comment (hiddeco) [reply pkg/bootstrap/git/gogit/gogit.go:196]: `isRemoteBranchNotFoundErr`?

Comment (stefanprodan) [reply pkg/bootstrap/git/gogit/gogit.go:196]: sounds good

**Comment by stefanprodan** [comment]
While testing this I found that if you run `bootstrap github/gitlab` for a repo then you change the repo name and you run it again, then the deploy key will not be set on 2nd repo and it will fail to pull.

**Comment by hiddeco** [comment]
@stefanprodan this behavior is also true for the current bootstrap logic, and is due to the presence of the secret in the cluster being the determining factor for the generation and/or upsert of the deploy key.

I see little room for workarounds, as I do not thing upserting the data from the secret in the cluster is good (security) practice, given anyone can have put anything in that secret.

**Comment by stefanprodan** [comment]
> I see little room for workarounds, as I do not think upserting the data from the secret in the cluster is good (security) practice, given anyone can have put anything in that secret.

We could check that there is no deploy key in the repo but there is a secret, if that's the case then bootstrap should fail with an error message that tells the user to either delete the secret or create the deploy key manually.

**Comment by stefanprodan** [comment]
When bootstraping a HTTPS repo with:

```
./bin/flux bootstrap git \
  --url=https://github.com/gitopsrun/git-test.git \
  --password=$GITHUB_TOKEN \
  --branch=main \
  --path=clusters/staging
► cloning branch "main" from Git repository "https://github.com/gitopsrun/git-test.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ committed sync manifests to "main" ("5a0611c25d7fd9e24327085003fecfbbb4348152")
► pushing component manifests to "https://github.com/gitopsrun/git-test.git"
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
✔ public key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDFa4pnuv3XDnUWxOpFX/j5WOfGUtaE9YxzF09II3yLZTrCRgqbkhj/RUAKvgqyO45+8mK7bFzaB1YR2VMIKf82PCHx8QapfDravskIxj2wSdqr/dkvtRbY5ySMzNgPoD/TEUVMhtHMNQyC6PmdvAyhTlQB1C1b9dlVgPsQPlq3zld6PFZGc0SR5NLnECk3ZTzzFCyKfzqDhsq+Y2aa+7cKKrVfg2LBwrkLuuuM5CyZNTTvD8qU2tZtQfSJbFGCIBfHXuhxGpO9dexAU9elpl75YUfMovUxQr+XGzh9bST5ywqdQEgVluI/jn4vcjE/rYE1NhM06n6sHS6wU+9YS2L1
Please give the key access to your repository: y█
► applying source secret "flux-system/flux-system"
✔ reconciled source secret
► generating sync manifests
✔ generated sync manifests
✔ committed sync manifests to "main" ("d9da28b945113d4e61ab778477c552c92f469a2d")
► pushing sync manifests to "https://github.com/gitopsrun/git-test.git"
► applying sync manifests
✔ applied sync manifests
✗ failed waiting for Kustomization: context deadline exceeded
```

Instead of generating a SSH key we should use basic auth.

**Comment by hiddeco** [comment]
> We could check that there is no deploy key in the repo but there is a secret, if that's the case then bootstrap should fail with an error message that tells the user to either delete the secret or create the deploy key manually.

This sounds nice in theory, but is not straight forward to implement in practice due to the Git provider implementation being a superset of the native Git one. As a result, the callback is only called _after_ the secret has been generated. I could theoretically also call it with the secret retrieved from the cluster _before_ generating, but there is nothing at present that provides context to the callback about the existence of the secret, making it a bit of a challenge without adding an obnoxious boolean or other flag type.

Any suggestions on a straight forward implementation that I may have overlooked? As I have been living in and looking at this code for too long, and may suffer from tunnel vision.

> Instead of generating a SSH key we should use basic auth.

This has been solved based on the details from our chat.

**Comment by stefanprodan** [comment]
I've tested bootstrap for GitHub and Git will all the flag combinations, took half a day. I didn't found any more issues. GG @hiddeco 🥇

**Comment by stealthybox** [comment]
Related #1165

**Comment by SomtochiAma** [comment]
The `--owner` flag description should highlight that it only accepts the actual path / textual path for GitLab groups. 

Great work @hiddeco 😁. It has been working okay so far.

**Comment by hiddeco** [comment]
From @SomtochiAma's testing:

```console
$ flux bootstrap gitlab --owner=root --repository=fleet-infra --branch=master --path=./staging-cluster/flux-system --hostname=https://custom --personal
```

returns

```console
✔ repository "https://https://example.cm/root/fleet-infra" reconciled
► cloning branch "master" from Git repository "https://https://example.com/root/fleet-infra.git"
✗ failed to clone repository: Get "https://https//example.com/root/fleet-infra.git/info/refs?service=git-upload-pack": dial tcp: lookup https on 192.168.0.1:53: no such host
```

but running without `https` in `--hostname`:

```console
► connecting to example.com
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x23a10e8]
goroutine 1 [running]:
github.com/fluxcd/go-git-providers/gitprovider/cache.(*cacheRoundtripper).RoundTrip(0xc000428120, 0xc00094c200, 0xc000428120, 0x0, 0x0)
	../go/pkg/mod/github.com/fluxcd/go-git-providers@v0.0.3/gitprovider/cache/httpcache.go:71 +0xc8
net/http.send(0xc00094c200, 0x2c3d680, 0xc000428120, 0x0, 0x0, 0x0, 0xc000428af8, 0x203000, 0x1, 0x0)
	/usr/local/go/src/net/http/client.go:251 +0x454
net/http.(*Client).send(0xc0008b4030, 0xc00094c200, 0x0, 0x0, 0
```

Likely related to https://github.com/fluxcd/go-git-providers/issues/55

**Review by stefanprodan — APPROVED** [review]
LGTM

Awesome work @hiddeco 🥇 🎖️ 🏅

**Pull request merged**